File: C++/ArrayString/20_numberOfSubarrayswithBoundedMax.cpp
#include <bits/stdc++.h>
using namespace std;
int subarrayWithMaxElement(vector<int> &nums, int k)
{
    int n = nums.size();
    int s = 0;
    int i = 0;
    while (i < n)
    {
        if (nums[i] > k)
        {
            i++;
            continue;
        }

        int count = 0;
        while (i < n and nums[i] <= k)
        {
            i++;
            count++;
        }
        s += (count * (count + 1)) / 2;
    }
    return s;
}

int numSubarrayBoundedMax(vector<int> &nums, int left, int right)
{
    if (nums.size() <= 0)
        return 0;

    return subarrayWithMaxElement(nums, right) - subarrayWithMaxElement(nums, left - 1);
}

File: Java/ArrayString/20_numberOfSubarrayswithBoundedMax.java
public int numSubarrayBoundedMax(int[] nums, int left, int right) {
        int n = nums.length, count =0;
        int dp[] = new int[n];
        int si = 0;
        for(int ei=0;ei<n;ei++){ 
            if(nums[ei] < left){
                dp[ei] = (ei-1 < 0) ? 0 : dp[ei-1];
            }
            else if(nums[ei] > right){
                dp[ei] = 0;
                si = ei+1;
            }
            else{
                dp[ei] = ei - si +1;
            }

        }
        
        for(int ele: dp) count+= ele;
        return count;
    }